### PR TITLE
lib/protocol: Don't block on Close (fixes #5794)

### LINKED ISSUE #5794: "Auto accept folders" doesn't work, deadlocks.
As of right now (1.2.0-rc.1) auto accepting new folders doesn't work. Instead Syncthing deadlocks when trying to add the folder. The two opposing routines are, one:

```
goroutine 31 [semacquire]:
sync.runtime_Semacquire(0xc000036748)
        /usr/local/go/src/runtime/sema.go:56 +0x39
sync.(*WaitGroup).Wait(0xc000036740)
        /usr/local/go/src/sync/waitgroup.go:130 +0x65
github.com/syncthing/syncthing/lib/model.(*model).handleAutoAccepts(0xc0001aa160, 0xb6e7244db96f5164, 0xffaeeb46980dd237, 0x27c86839ef5695b0, 0xc244c124fec3ef6f, 0xc0000cc8c0, 0x2, 0xc00014c510, 0x1, 0x1, ...)
        /Users/jb/dev/github.com/syncthing/syncthing/lib/model/model.go:1350 +0xa88
github.com/syncthing/syncthing/lib/model.(*model).ClusterConfig(0xc0001aa160, 0xb6e7244db96f5164, 0xffaeeb46980dd237, 0x27c86839ef5695b0, 0xc244c124fec3ef6f, 0xc00039e080, 0x2, 0x2)
        /Users/jb/dev/github.com/syncthing/syncthing/lib/model/model.go:1053 +0x406
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).dispatcherLoop(0xc0000fc410, 0x0, 0x0)
        /Users/jb/dev/github.com/syncthing/syncthing/lib/protocol/protocol.go:407 +0x116e
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).Start.func1(0xc0000fc410)
        /Users/jb/dev/github.com/syncthing/syncthing/lib/protocol/protocol.go:255 +0x2b
created by github.com/syncthing/syncthing/lib/protocol.(*rawConnection).Start
        /Users/jb/dev/github.com/syncthing/syncthing/lib/protocol/protocol.go:254 +0x65
```

and two:

```
goroutine 181 [chan receive]:
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).internalClose.func1()
        /Users/jb/dev/github.com/syncthing/syncthing/lib/protocol/protocol.go:903 +0x1de
sync.(*Once).Do(0xc0000fc4b8, 0xc001f04400)
        /usr/local/go/src/sync/once.go:44 +0xb3
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).internalClose(0xc0000fc410, 0x4c3d2a0, 0xc00023a3a0)
        /Users/jb/dev/github.com/syncthing/syncthing/lib/protocol/protocol.go:890 +0x6f
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).Close(0xc0000fc410, 0x4c3d2a0, 0xc00023a3a0)
        /Users/jb/dev/github.com/syncthing/syncthing/lib/protocol/protocol.go:885 +0x96
github.com/syncthing/syncthing/lib/connections.completeConn.Close(0x4c58320, 0xc000296a80, 0x3, 0xa, 0x4c5b940, 0xc00025c3c0, 0x4c3d2a0, 0xc00023a3a0)
        /Users/jb/dev/github.com/syncthing/syncthing/lib/connections/structs.go:43 +0x45
github.com/syncthing/syncthing/lib/model.(*model).closeConns(0xc0001aa160, 0xc00043c400, 0x2, 0x2, 0x4c3d2a0, 0xc00023a3a0)
        /Users/jb/dev/github.com/syncthing/syncthing/lib/model/model.go:1440 +0x23a
github.com/syncthing/syncthing/lib/model.(*model).startFolderLocked(0xc0001aa160, 0xc000226020, 0x7, 0x0, 0x0, 0x0, 0xc000036330, 0x9, 0x0, 0xc001ed2480, ...)
        /Users/jb/dev/github.com/syncthing/syncthing/lib/model/model.go:285 +0x41e
github.com/syncthing/syncthing/lib/model.(*model).StartFolder(0xc0001aa160, 0xc000226020, 0x7)
        /Users/jb/dev/github.com/syncthing/syncthing/lib/model/model.go:251 +0x14a
github.com/syncthing/syncthing/lib/model.(*model).CommitConfiguration(0xc0001aa160, 0x1d, 0x5292c28, 0x0, 0x0, 0xc001ee8500, 0x5, 0x5, 0x1, 0xc0000ccd30, ...)
        /Users/jb/dev/github.com/syncthing/syncthing/lib/model/model.go:2478 +0x563
github.com/syncthing/syncthing/lib/config.(*wrapper).notifyListener(0xc0000a8800, 0x4c4d580, 0xc0001aa160, 0x1d, 0x5292c28, 0x0, 0x0, 0xc001ee8500, 0x5, 0x5, ...)
        /Users/jb/dev/github.com/syncthing/syncthing/lib/config/wrapper.go:239 +0x154
github.com/syncthing/syncthing/lib/config.(*wrapper).notifyListeners.func1(0xc0000a8800, 0xc000297500, 0xc000297880, 0x4c4e400, 0xc000036740, 0x4c4d580, 0xc0001aa160)
        /Users/jb/dev/github.com/syncthing/syncthing/lib/config/wrapper.go:230 +0x121
created by github.com/syncthing/syncthing/lib/config.(*wrapper).notifyListeners
        /Users/jb/dev/github.com/syncthing/syncthing/lib/config/wrapper.go:229 +0x1bf
```

In essense:

1. clusterconfig receiver updates config and waits for config to commit
2. config committer closes connections as part of commit, and waits for that
3. connection that is being closed waits for the dispatcher loop to exit
4. the dispatcher loop is busy calling clusterconfig in step one

(full trace attached)

Pretty sure this is new with the protocol close handling changes in 1.2.0-rc.1, so an RC issue.

[trace.txt](https://github.com/syncthing/syncthing/files/3290291/trace.txt)

### PR DESCRIPTION
**Pull request by imsodin**
https://github.com/syncthing/syncthing/issues/5794

### CODE REVIEW

**Comment by calmh** [comment]
LGTM. Windows builder is down, I'm moving it to a new build server (moar power), once that's up and green it's a go.

File: lib/protocol/protocol_test.go [hunk 1]
```diff
@@ -813,3 +813,20 @@ func TestClusterConfigAfterClose(t *testing.T) {
 		t.Fatal("timed out before Cluster Config returned")
 	}
 }
+
+func TestDispatcherToCloseDeadlock(t *testing.T) {
```
Comment (calmh) [inline lib/protocol/protocol_test.go]: // Verify that we don't deadlock when calling Close() from within one of the model callbacks (ClusterConfig).

**Comment by calmh** [comment]
Hope you don't mind me throwing comments and stuff into the code without asking, it's roughly the same amount of effort for me to do it instead of asking for it and it avoids a review cycle...

**Comment by imsodin** [comment]
Sure, I definitely don't mind.

**Pull request merged**